### ISSUE TITLE
MySql Connector throw an exception with a big concurrency volume

### ISSUE THREAD

**Issue by Kagamine** [mysql-net/MySqlConnector#196]
![image](https://cloud.githubusercontent.com/assets/2216750/24034704/64e42ade-0b2d-11e7-92c6-7d9046c9cf09.png)

**Comment by Kagamine** [comment]
I'm using mysql on azure(5.7) which provided by Microsoft China Cloud Innovation Center officially. Disabled `reset connection`.

**Comment by Kagamine** [comment]
Hi, @bgrainger, @edwinsongmsft is the Principal Program Manager of Microsoft MySQL on Azure team. If you want to know what is the differences between normal MySQL and MySQL on Azure, you could communicate with him.

**Comment by edwinsongmsft** [comment]
Hi @bgrainger, as we did not support the auto reconnection, could u catch the exception and reconnect the database again? BR, Edwin

**Comment by bgrainger** [comment]
From the exception message, it seems like `MySqlSession.TryPingAsync` is being called on a session that is associated with a socket that already has an in-flight operation. I'm not exactly sure how to get into this situation; do you have more code that reproduces the problem that you can share?

**Comment by Kagamine** [comment]
it occurred in a big concurrency volume. There was no particular code.

**Comment by caleblloyd** [comment]
Possible duplicate. This one uses MariaDB.

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/221

**Comment by bgrainger** [comment]
"There is already an open DataReader associated with this Connection" (from the linked issue) could indicate incorrect use of the library. If that is the root cause, then perhaps it's possible that it could also cause the "An asynchronous socket operation is already in progress" exception?

If so, a good troubleshooting step might be to add explicit checks in public APIs of MySqlConnector to detect this situation and throw a more helpful exception earlier.

OTOH, there could be a bug in this library that's causing both exceptions. If we add that code to throw the new exception and it's not getting thrown, that would discount the "incorrect use" hypothesis.

**Comment by caleblloyd** [comment]
Could this be coming from improperly disposed readers?

Do we force the reader to dispose when a connection is disposed? We might need to check the logic there. Since read buffering was added, we may need to add code to clear the read buffer as well.

**Comment by bgrainger** [comment]
Good call: `MySqlConnection` has `HasActiveReader`, but `DoClose` doesn't check it. (It will check if there's an active transaction and dispose (rollback) that.)

Disposing any active reader in `MySqlConnection.DoClose` may well work around this bug. (And once cancellation is actually correctly implemented, it may end up being more efficient to cancel the reader, rather than just read all its rows to the end.) And hopefully it wouldn't be too hard to reproduce this problem.

**Comment by caleblloyd** [comment]
> Since read buffering was added, we may need to add code to clear the read buffer as well.

Adding Reader Disposal to `MySqlConnection.DoClose` should take care of reading through anything that is buffered as well.

**Comment by bgrainger** [comment]
Should be fixed in [0.14.1](https://www.nuget.org/packages/MySqlConnector/0.14.1).

**Comment by Kagamine** [comment]
@bgrainger There was another exception occurred with a big concurrency volume

![image](https://cloud.githubusercontent.com/assets/2216750/24189273/3f8efe8a-0f1f-11e7-9a18-165f80f309e5.png)

**Comment by bgrainger** [comment]
This last exception is a duplicate of #162.

**Comment by Kagamine** [comment]
@bgrainger When will you fix #162? Our customer case was blocked by this.

**Comment by bgrainger** [comment]
@Kagamine Is it possible that your customer uses cancellation in any way, i.e., calls `Async` overloads with something other than `CancellationToken.None`? My working hypothesis is that this is related to currently-unimplemented cancellation support #3 but so far I have been unable to obtain a consistent repro to prove/disprove that.
